Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift

```diff
@@ -0,0 +1,1064 @@
+                    let delegate = self.delegate
+                    self.delegate = nil
+
+                    return delegate
+
+                case .none:
+                    return nil
+                }
+            }
+
+            delegate?.didTerminate()
+        }
+
+        @inlinable
+        /* fileprivate */ internal func yield<S: Sequence>(_ sequence: S) -> Source.YieldResult where S.Element == Element {
+            self.lock.withLock {
+                let action = self.stateMachine.yield(sequence)
+
+                switch action {
+                case .returnProduceMore:
+                    return .produceMore
+
+                case .returnStopProducing:
+                    return .stopProducing
+
+                case .returnDropped:
+                    return .dropped
+
+                case .resumeContinuationAndReturnProduceMore(let continuation, let element):
+                    // It is safe to resume the continuation while holding the lock
+                    // since the task will get enqueued on its executor and the resume method
+                    // is returning immediately
+                    continuation.resume(returning: element)
+
+                    return .produceMore
+
+                case .resumeContinuationAndReturnStopProducing(let continuation, let element):
+                    // It is safe to resume the continuation while holding the lock
+                    // since the task will get enqueued on its executor and the resume method
+                    // is returning immediately
+                    continuation.resume(returning: element)
+
+                    return .stopProducing
+                }
+            }
+        }
+
+        @inlinable
+        /* fileprivate */ internal func finish() {
+            let delegate: Delegate? = self.lock.withLock {
+                let action = self.stateMachine.finish()
+
+                switch action {
+                case .callDidTerminate:
+                    let delegate = self.delegate
+                    self.delegate = nil
+
+                    return delegate
+
+                case .resumeContinuationWithNilAndCallDidTerminate(let continuation):
+                    let delegate = self.delegate
+                    self.delegate = nil
+
+                    // It is safe to resume the continuation while holding the lock
+                    // since the task will get enqueued on its executor and the resume method
+                    // is returning immediately
+                    continuation.resume(returning: nil)
+
+                    return delegate
+
+                case .none:
+                    return nil
+                }
+            }
+
+            delegate?.didTerminate()
+        }
+
+        @inlinable
+        /* fileprivate */ internal func next() async -> Element? {
+            // We are locking manually here since we want to hold the lock
+            // across the suspension point.
```

Comment:
this comment is a bit of a lie; you cannot hold across a suspension; what you're doing though is fine, since you unlock still in the SYNCHRONOUS piece of the withContinuation -- that is fine; it does not actually hold over the suspension point, due to how continuations work

Reply:
(best to remove this comment; the one below is correct tho)